README
"proc" is a colored top like utility which combines the features
of many other 'top'-like tools, but has more integrated features.
These include:

   * History - you can scroll backwards to see the state of the system
     at prior times.
   * More commands.
   * More detail on the commands.
   * Ability to see absolute or relative counter values.
   * Use of color.

Command Summary
---------------
abs        Show absolute values.
agg        Aggregate thread cpu into parent.
clear      Clear counter values.
count      Start counting mode.
cpu        Show CPU details.
delay      Set refresh delay.
delta      Show delta values.
disk       Show disk statistics..
files      Show open file information
flush      Flush open file from disk cache
graph      Common graphs (loadavg, dirty memory, cached memory, etc)
grep       Filter process output
help       Display help
hide       Hide a column
icmp       Show IP/ICMP statistics
ifconfig   Show network interface stats
ip         Show IP/ICMP statistics
interrupts Show /proc/interrupts
log        [on | off] Enable/disable logging
meminfo    Memory info
netstat    Show socket information
order      [[-]nprstuc] Normal; Proc; RSS; Size; Time; User; CPU
proc       Show specific proc
ps         Display process mode
quit       Exit from tool
range      Show numbers in units/ranges
set        Set options/limits.
show       Show a hidden column
snap       Dump snapshots to file to see history
softirqs   Show /proc/softirqs
tcp        Show TCP statistics
thread     Toggle thread mode
udp        Show UDP statistics
units      Show units on numbers
vmstat     Show vmstats
?          Display help
!          Run a unix command repeatedly

Ctrl-B - page/back; Ctrl-F - page/fwd; Ctrl-R - reset paging
Ctrl-L - redraw screen

When viewing the statistics screens, you may find it useful to
select "delta" mode if the statistics are monotonically increasing
counters. This makes it easier to spot which ones are changing and
how rapidly.

You can use the "clear" command to reset the onscreen counters to
make it easier to start a performance monitoring session or to just
watch, removing the noise from the history prior to the viewing session.

Header information
------------------
The header shows a number of fields, all incrementals from
the prior sample (typically 1s sample).

last pid:       Shows the last pid created. If this is jumping in big
                leaps then a fair amount of forking is going on. If
		the new processes are not visible in the ps listing, then
		it could be short lived processes (and %sys time
		will likely be higher to handle the expense of processing
		the forks/execs).

in:             Interrupts. High levels of interrupts could be due to
                I/O or network activity.

load avg:       1, 5 and 15 min load average.

Boot:           Time when system last booted.

(Reverse) HH:MM:SS
                When in rewind mode, the time of the sample (via
		Ctrl-B/Ctrl-F)

HH:MM:SS        Current time. Useful to watch in case no other activity.

CPU:            Summary of the CPU characteristics (from /proc/cpuinfo)
                If hyperthreading enabled, then "(HT)" will appear to
		show the number of cores may be doubled up.

proc:           Number of processes currently running or stopped or zombie.
thr:            Number of threads.

RAM:            Amount of memory in the system
RSS:            Amount of resident working set for processes
Free:           How much memory is free. Should be close to zero on a
                normal system. Kernel will try and use memory for process
		heaps or buffer cache. High free figure can indicate
		large processes have terminated, or large files have
		been deleted, or the memory hasnt been used since
		bootup.
Dirty:          How many dirty pages in memory which need to be committed to
                disk. High figure indicates active I/O. Can be high when
		doing large writes to slow devices (eg flash memory).
Swap:           How much swap space is in use.
Free:           How much swap space is available.


Absolute / Delta display mode
-----------------------------

Some commands simply show the pure data from /proc (eg meminfo and
vmstats). Many of these values are incrementing counters, and after a long
period of uptime, the number of digits is large enough that even
reading the numbers is hard work. Mentally determining what is
going on is equally difficult. For these reports, a delta is shown
next to the number to show if it is incrementing or decrementing
since the last display. 

To make it easier to read, a historical graph of recent activity is
shown (typically around 2mins of history).

The counters can be set to show absolute values or delta values.
Delta values can be useful to show the count by 'zeroing' the value
and watching the values rise.

In addition the values can be displayed as pure numbers or as
K/M/G to indicate thousands, millions, billions.

The commands to use are:

	* abs
	* delta
	* ranged
	* clear

Detailed command summary
------------------------
Process information
Network information
disk         Shows the details from /proc/diskstats. Graphs are shown
             for each numeric field. It may be better to use the 'delta'
	     mode, otherwise you will only see big numbers incrementing.

CPU	     Shows the details from /proc/cpuinfo and the current MHz
             speed of the cpu.

grep [!]expr Use this command to restrict the view to lines which match
             or (!) do not match the patern. At present, the pattern
	     is a literal match (no regexp support). Prefix the pattern
	     with "!" to do a negative match. The expression is used
	     to match columns, rather than specific lines. Not all
	     views support the feature.

meminfo	     Show /proc/meminfo. Only non-zero values are displayed.

netstat      Show current socket information. Additional details include
             the owning process (if it can be determined). Given
	     that a socket can be shared amongst processes, only one
	     process is shown (use lsof utility or examine /proc/pid/fd
	     to find the duplicate owners).

	     The display is a multiline display. The first line shows
	     the owner of the socket, the rcv/snd data queue, connection
	     state and local/remote address, in numeric format. Subsequent
	     lines show the local/remote address as names. If the owning
	     process can be determined, then the PID and process name is shown.

	     If no pid/command line is shown, then we are not running as root
	     and cannot determine the owner.

	     You can use the "grep" command to restrict the view to
	     matching lines or non-matching lines. (Prefix grep
	     pattern with "!", e.g. "grep !LISTEN").

	     If the socket is new (less than about 20s), then the owner
	     is displayed in reverse video.

Open files / cache usage
proc <pid>   Show details for a process, including /proc/pid/status,
             /proc/pid/stack and /proc/pid/syscall.

softirqs     Show /proc/softirqs. /proc/softirqs is used to indicate
             deferred interrupt level processing (which is painful to
	     SMP configurations). Instead the interrupt routine is
	     deferred until it can be enabled whilst other interrupts
	     can be allowed. Its normal to see these on a well running
	     kernel. Higher numbers indicate more interrupt processing,
	     but also better perceived performance.

vmstat       Show /proc/vmstat. There are lots of VM statistics and they
             vary from kernel to kernel. The absolute value and a delta
	     from the prior value are shown, along with a graph showing
	     recent history. You can change the display mode to
	     abs/delta mode and reset the counters.

"proc" includes 20 minutes of buffered history - you can cycle
back in time, to take a closer look at something which gained
your attention.

Other commands
--------------
grep [<pattern>]      Only show lines matching the pattern. Specify no
                      pattern to restore the full list.

History Mode
------------

Use the following keys to cycle through the history

<Ctrl-B>    Move back 1s of time
<Ctrl-F>    Move forward 1s of time
<Ctrl-R>    Reset to 'realtime'

Whilst in history mode, the realtime clock at the top right of the
screen will adjust to the appropriate time sample.

Type 'help' whilst running to see the interactive commands available.

Environment Variables
---------------------

PROC_MAX_SYMS      Number of shared memory symbol slots to reserve.
                   On later kernels, or kernels with lots of devices,
		   the default values may be too small, so you may want
		   to increase this, if "proc" complains of too few
		   values.
		   Default: 2048

PROC_MAX_TICKS     Number of history slots to reserve per symbol. The
                   amount of memory is directly proportional to this
		   number. Each number corresponds to 1s of history.
		   (1024 is rougly 20mins and about 300+MB of memory).

Files
-----
"proc" uses shared memory to communicate between the collector/monitor.
You will see 'procmon' in the process table for the collector. Only
one process will run for the system - the collector is shared amongst
the proc clients.

Files are stored in /tmp/proc (or $TMPDIR/proc is $TMPDIR is
defined).

Security Considerations
-----------------------
"proc" does not need to run as root, but some information may be
missing if it does not run as root. E.g. socket owning information
is one example.

Running proc as root, on your own system is perfectly reasonable, but
be careful of running as root on a public or limited access system.
proc can be used to kill process, spawn process, and writes to files
in the /tmp/proc directory.


Changes
Thu Jul 23 22:50:15 2015  fox

      58* .gitignore
          common/commands.c
          linux/graphs.c
          linux/proc.c
          linux/sigdesc.h
          makefile
          scripts/stackmon.pl
          Add pids to graph. Attempt to fix infinite loop where we dont
	  read kbd on ^B; add support for ^N/^P

Sat Aug 30 19:35:23 2014  fox

      57* .mk
          linux/monitor.c
          linux/proc.c
          linux/sigdesc.h
          makefile
          mk/common.mk
          mk/targets.mk
          Refresh the build to avoid cache username() issue.

Thu Oct 11 21:58:46 2012  fox

      56* monitor.c: Default to /var/tmp.

Tue Jul 10 20:20:31 2012  fox

      55* proc.c: Fix for RSS being out by *4.

Thu Jun 21 16:06:45 2012  fox

      54* makefile: Build 32 + 64 bit binaries. Hard code/internalise
          and remove dependency on ncurses.

Tue Oct 11 21:37:49 2011  fox

      53* monitor.c: Really fix the last-proc pctcpu calc so it works properly.
          More sillyness in thread vs proc sort/searching.

Mon Oct 10 23:22:20 2011  fox

      52* monitor.c: Fix stupid last-proc vs last-task sort/search which
          caused stupid %CPU for processes.

Sun Oct  9 10:25:14 2011  fox

      51* monitor.c: Better atomic updating.

Tue Sep 20 21:18:20 2011  fox

      50* monitor.c: Avoid pathological condition with procmon dying and us
          spinning the cpu.

Wed Sep 14 21:40:17 2011  fox

      49* proc.c, monitor.c: Better handling of corrupt/missing pmap files.

Sat Sep 10 22:36:43 2011  fox

      48* Add interrupts command, along with graphs.

Fri Sep  9 21:28:28 2011  fox

      47* proc.c: Fix EVENT field for 64-bit wchan

      46* proc.c: Make sure we turn off graph mode on ^C/quit.

Thu Sep  8 22:58:47 2011  fox

      45* protocol.c, graph.c: Display minigraphs as delta values.

Wed Sep  7 23:01:56 2011  fox

      44* Add: ip, icmp, udp, tcp - stats screens.

Sun Sep  4 17:48:44 2011  fox

      43* proc.c: Do binary search to find original pids, instead of linear.

      42* proc.c: Better handling of procmon terminating/restarting on us.

      41* monitor.c: Avoid reading /proc/pid/syscall if we are not root
          or its not our uid (performance).

Sat Sep  3 22:14:10 2011  fox

      40* proc.c, monitor.c: Add stale procmon detector.

Fri Sep  2 22:39:23 2011  fox

      39* monitor.c: Avoid pointless reads of /proc/self/syscall.

      38* proc.c: Avoid double (pointless) read on /proc/pid/cmdline.

Thu Sep  1 22:54:55 2011  fox

      37* monitor.c: When looking at a TIME_WAIT socket, try and maintain
          the userid of the time wait original owner.

      36* netstat.c: Add persistent port caching.

Wed Aug 31 22:08:15 2011  fox

      35* netstat.c: Avoid potential port storm.

Tue Aug 30 21:09:39 2011  fox

      34* netstat.c: Sort sockets into new/time order. Highlight new sockets.
          Add support for grep on state.

      33* netstat.c: Add port name cache.

Mon Aug 29 20:34:22 2011  fox

      32* netstat.c: Various fixes/enhancements. Still work in progress.

Sun Aug 28 23:30:50 2011  fox

      31* netstat.c: Add dynamic 'netstat' command.

Sat Aug 27 10:40:41 2011  fox

      30* netstat.c: Rename to ifconfig. Add graphs support

      29* proc.c: Implement ^C handler to clear the graphics.

      28* monitor.c: Handle auto-resizing of the mmap when we run out
          of symbols.

Sun Aug 21 22:08:10 2011  fox

      27* Add graphs to the 'disk' command.

Fri Aug 12 20:27:15 2011  fox

      26* Fix pid and cpu times caused by using %02d format in mmap file.

Thu Aug 11 10:22:08 2011  fox

      25* Archive: 1.0003

      24* proc.c: Fix %CPU to be a moving average 80/20.

      23* proc.c, monitor.c: Implement locking protocol to avoid time changing
          under us on ps screen.

      22* monitor.c: Fix realpath() in finding procmon.

Wed Aug 10 10:59:50 2011  fox

      21* Archive: 1.0002

Tue Aug  9 10:49:37 2011  fox

      20* Lots of fixes. 'ps' mode now works from history, as does
          the header. 'help all' now displays the README.

Sun Aug  7 11:23:05 2011  fox

      19* Get proc mem size delta to work again.

      18* Add THR column.

      17* Add historic proc access.

Wed Aug  3 23:30:26 2011  fox

      16* add support for abs/delta/clear modes (works for softirq,
          but nothing else just yet).

Mon Aug  1 22:44:30 2011  fox

      15* proc.c, monitor.c: Better time history and sync.

      14* softirqs.c: Add softirqs support.

Fri Jul 29 21:13:11 2011  fox

      13* add background processing and history navigation.

Mon Jul 25 22:07:05 2011  fox

      12* add meminfo command.

Sun Jul 24 22:03:52 2011  fox

      11* Add 'disk' command and 'netstat' command.

Fri Jul 22 21:23:40 2011  fox

      10* proc.c: Add 'cpu' command.

Wed Jul 20 22:24:44 2011  fox

       9* proc.c: Load $HOME/.proc on startup.

       8* comamnds.c: Show the column state on hide/show if no args.

Sat Jan  2 22:27:05 2010  fox

       7* linux/proc.c: Add SYSCALL column

       6* linux/proc.c: Use /proc/pid/wchan

Tue Sep 30 20:16:47 2008  fox

       5* commands.c: Add flush command.

Fri Sep 19 23:20:07 2008  fox

       4* proc.c: Correct the stats for threads.

Thu Sep 18 21:27:21 2008  fox

       3* fix RSS overflow (maybe)

       2* Show thread pids in white, vs green for real procs.

       1* proc.c: Support later kernel /proc/pid/task threads.



proc.pl
#! /usr/bin/perl

# $Header:$

use strict;
use warnings;

use File::Basename;
use FileHandle;
use Getopt::Long;
use IO::File;
use POSIX;

my %stacks;
my $stk_no;
my %cmdline;
my $cmd_no;

my $do_syscall = -f "/proc/self/syscall";

#######################################################################
#   Command line switches.                                            #
#######################################################################
my %opts = (
	n => 1,
	);

sub main
{
        Getopt::Long::Configure('require_order');
        Getopt::Long::Configure('no_ignore_case');
        usage() unless GetOptions(\%opts,
                'help',
		'n=s',
                );

        usage() if $opts{help};

        %stacks = ();
        $stk_no = 0;

	for (my $i = 0; $i < $opts{n}; $i++) {
	        my $dh;
	        opendir($dh, "/proc");
	        while (my $name = readdir($dh)) {
	                next if $name !~ /^[0-9]/;
	                my $str = dump_proc("/proc/$name", 1);
			print $str;
	        }
	        closedir($dh);

	        foreach my $s (keys(%cmdline)) {
	                print $s, "\n";
	        }
	        foreach my $s (keys(%stacks)) {
	                print $s, "\n";
	        }
		print "===\n";
		last if $i + 1 >= $opts{n};
		sleep(1);
	}
}
sub dump_proc
{       my $p = shift;
        my $root_dir = shift;

        my $pid = basename($p);

        my $fh;

        print "=$pid\n";
        my $str = '';
        my $str1 = read_file("$p/stat");
	if ($root_dir) {
	        $str .= join(" ", (split(" ", $str1))[
	#               0, #  readone(&pid);
	                1, #  readstr(tcomm);
	                2, #  readchar(&state);
	                3, #  readone(&ppid);
	                4, #  readone(&pgid);
	                5, #  readone(&sid);
	                6, #  readone(&tty_nr);
	                7, #  readone(&tty_pgrp);
	                8, #  readone(&flags);
	                9, #  readone(&min_flt);
	                10, #  readone(&cmin_flt);
	                11, #  readone(&maj_flt);
	                12, #  readone(&cmaj_flt);
	                13, #  readone(&utime);
	                14, #  readone(&stimev);
	                15, #  readone(&cutime);
	                16, #  readone(&cstime);
	                17, #  readone(&priority);
	                18, #  readone(&nicev);
	                19, #  readone(&num_threads);
	                20, #  readone(&it_real_value);
	                21, #  readunsigned(&start_time);
	                22, #  readone(&vsize);
	                23, #  readone(&rss);
	                24, #  readone(&rsslim);
	#               25, #  readone(&start_code);
	#               26, #  readone(&end_code);
	#               27, #  readone(&start_stack);
	                28, #  readone(&esp);
	                29, #  readone(&eip);
	                30, #  readone(&pending);
	                31, #  readone(&blocked);
	                32, #  readone(&sigign);
	                33, #  readone(&sigcatch);
	                34, #  readone(&wchan);
	#               35, #  readone(&zero1);
	#               36, #  readone(&zero2);
	#               37, #  readone(&exit_signal);
	                38, #  readone(&cpu);
	                39, #  readone(&rt_priority);
	                40, #  readone(&policy);
	                ]);
	} else {
	        $str .= join(" ", (split(" ", $str1))[
	#               0, #  readone(&pid);
	#               1, #  readstr(tcomm);
	                2, #  readchar(&state);
	#               3, #  readone(&ppid);
	#               4, #  readone(&pgid);
	#               5, #  readone(&sid);
	#               6, #  readone(&tty_nr);
	#               7, #  readone(&tty_pgrp);
	                8, #  readone(&flags);
	                9, #  readone(&min_flt);
	                10, #  readone(&cmin_flt);
	                11, #  readone(&maj_flt);
	                12, #  readone(&cmaj_flt);
	                13, #  readone(&utime);
	                14, #  readone(&stimev);
	#               15, #  readone(&cutime);
	#               16, #  readone(&cstime);
	                17, #  readone(&priority);
	                18, #  readone(&nicev);
	#               19, #  readone(&num_threads);
	                20, #  readone(&it_real_value);
	                21, #  readunsigned(&start_time);
	#               22, #  readone(&vsize);
	#               23, #  readone(&rss);
	#               24, #  readone(&rsslim);
	#               25, #  readone(&start_code);
	#               26, #  readone(&end_code);
	#               27, #  readone(&start_stack);
	                28, #  readone(&esp);
	                29, #  readone(&eip);
	                30, #  readone(&pending);
	                31, #  readone(&blocked);
	                32, #  readone(&sigign);
	                33, #  readone(&sigcatch);
	                34, #  readone(&wchan);
	#               35, #  readone(&zero1);
	#               36, #  readone(&zero2);
	#               37, #  readone(&exit_signal);
	                38, #  readone(&cpu);
	                39, #  readone(&rt_priority);
	                40, #  readone(&policy);
	                ]);
	}
        $str .= "\n";

#       $str .= read_file("$p/wchan");
        if ($root_dir) {
		$str .= "io: ";
                foreach my $s (split("\n", read_file("$p/io"))) {
			$s =~ s/^.*: //;
			$str .= "$s ";
		}
		$str .= "\n";
                my $cmdline = read_file("$p/cmdline");
                $cmdline =~ s/\0/ /g;
                if (!defined($cmdline{$cmdline})) {
                        $cmdline{$cmdline} = $cmd_no++;
                }
                $str .= "cmd$cmdline{$cmdline}\n";
        }
        if ($do_syscall) {
                $str .= read_file("$p/syscall");
        }
#       my $lnk = readlink("$p/exe") || "";

        $fh = new FileHandle("$p/stack");
        if ($fh) {
                my $s = '';
                while (<$fh>) {
                        last if /] 0xffff/;
                        $s .= $_;
                }
                if (!defined($stacks{$s})) {
                        $stacks{$s} = $stk_no++;
                }
                $str .= "stack=$stacks{$s}\n";
        }
        if ($root_dir) {
	        foreach my $p1 (glob("$p/task/[0-9]*")) {
	                next if basename($p1) == $pid;
	                $str .= dump_proc($p1, 0);
	        }
	}
	return $str;
}
sub read_file
{       my $fname = shift;

        my $fh;
        return "" if !sysopen($fh, $fname, 0);
        my $str;
        sysread($fh, $str, 2048);
       $str =~ s/18446744073709551615/-1/g;
        return $str;
}
#######################################################################
#   Print out command line usage.                                     #
#######################################################################
sub usage
{
        print <<EOF;
Some help...
Usage:

Switches:

EOF

        exit(1);
}

main();